### ISSUE TITLE
composerModel API: return multiple outputs from forward

### ISSUE THREAD

**Issue by cmlakhan** [mosaicml/composer#1918]
I am training a bert-like model in which the output of the model returns two vectors a class vector and a sequence vector.  The class vector is put through a separate linear head and then I want to calculate a total loss summing up the masked loss from Bert and the loss for the linear head.  Looking at the documentation I am unclear how I would do such a thing using the ComposerModel API.  It seems the output of forward is a class called output, but what do I do if I have two outputs?  Also I would like to log separate  metrics for each output e.g. accuracy and AUC for each of the tasks.  Is this sort of thing possible in Composer?  I have all of the logic written as a PyTorch lightning module if it would help to debug further.  I didn't see any multi-task learning examples so maybe this isn't possible with Composer?

**Comment by dakinggg** [comment]
Hi @cmlakhan, when defining a `ComposerModel`, you control your forward function and loss function (see the docs for more detail https://docs.mosaicml.com/en/latest/composer_model.html). The output of forward is just passed to your loss function, so you should be able to define how it gets interpreted to compute the loss.

**Comment by cmlakhan** [comment]
What would be the analogue to outputs in the loss function if I return 2 tensors in forward?

class ResNet18(ComposerModel):

    def __init__(self):
        super().__init__()
        self.model = torchvision.models.resnet18()

    def forward(self, batch): # batch is the output of the dataloader
        # specify how batches are passed through the model
        inputs, _ = batch
        return self.model(inputs)

    def loss(self, outputs, batch):
        # pass batches and `forward` outputs to the loss
        _, targets = batch
        return F.cross_entropy(outputs, targets)

**Comment by dakinggg** [comment]
Sorry, I don't quite understand. `outputs` in the loss function is just exactly what you return from `forward`.

**Comment by cmlakhan** [comment]
sorry if I am being dense, but if I have a model like this

```
def __init__(self):
    super().__init__()
    self.model = model()

def forward(self, batch): # batch is the output of the dataloader
    # specify how batches are passed through the model
    inputs, _ = batch
    vec1, vec2 = self.model(inputs)
    return vec1, vec2
```

how would I access ```vec1``` and ```vec2``` in my loss function would it be ```output[0]``` and ```output[1]``` or something else?

**Comment by cmlakhan** [comment]
So is the correct way to do things?  sorry just trying to make sure I understand the function inputs for loss


```

class model(ComposerModel):

def __init__(self):
    super().__init__()
    self.model = model()

def forward(self, batch): # batch is the output of the dataloader
    # specify how batches are passed through the model
    inputs,_ ,_ = batch
    vec1, vec2 = self.model(inputs)
    return vec1, vec2

def loss(self, vec1, vec2, batch):
    # pass batches and `forward` outputs to the loss
    _, targets1, targets2 = batch
loss1 = F.cross_entropy(vec1, targets1)    
loss2 = F.cross_entropy(vec2, targets2) 
return loss1+loss2
```

**Comment by dakinggg** [comment]
I think you want something like, let me know if that works for you!

```python
class Model(ComposerModel):
    def __init__(self):
        super().__init__()
        self.model = model()
    
    def forward(self, batch): # batch is the output of the dataloader
        # specify how batches are passed through the model
        inputs,_ ,_ = batch
        vec1, vec2 = self.model(inputs)
        return vec1, vec2
    
    def loss(self, outputs, batch):
        # pass batches and `forward` outputs to the loss
        _, targets1, targets2 = batch
        loss1 = F.cross_entropy(outputs[0], targets1)    
        loss2 = F.cross_entropy(outputs[1], targets2) 
        return loss1+loss2
```

**Comment by cmlakhan** [comment]
Thanks, will try this out and let you know

**Comment by mvpatel2000** [comment]
@cmlakhan wanted to check to see if you were able to resolve this?

**Comment by cmlakhan** [comment]
You can close it and I will reopen when if I have issues.

**Comment by mvpatel2000** [comment]
Got it. Please feel free to reach out if you come back to this!
